src/installer/corehost/cli/apphost/bundle/header.h

```diff
@@ -27,7 +32,31 @@ namespace bundle
         uint32_t minor_version;
         int32_t num_embedded_files;
 
-        bool is_valid() const;
+        bool is_valid(bool exact_match = false) const;
+    };
+#pragma pack(pop)
+
+    enum header_flags_t : uint64_t
+    {
+        none = 0,
+        needs_extraction = 1,
+        netcoreapp3_compat_mode = 2
```

Comment:
To be honest I don't like the compat mode flag. Can't we deduce it from the missing V2 header instead? I find it really weird that we would have the new bundle, but try to mimic the old behavior.
Basically - I would expact behavior like this:
* `3.*` SDK will always use `3.*` `apphost` and also V1 header only
* `5.*` SDK building `netcoreapp3.*` app will use `3.*` `apphost` and V1 header only
* `5.*` SDK building `net5` app will always use `5.*` `apphost` and V2 header

The presence of the flag means that there will be a combination of `5.*` apphost used on `netcoreapp3.*` - I find that wrong.

Reply:
I agree that apphosts will always match the TFM, exactly as you've noted above, regardless of the SDK. 

The `netcoreapp3_compat_mode` flag is set on a .net5 app, which chooses to build single-file apps in .netcore3.x compat mode, through the [`IncludeAllContentInSingleFile`](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#optional-settings) setting. 

This means: 
* All files are bundled into the app; some of them will be extracted to disk.
* `AppContext.BaseDirectory` is set to the extraction directory. Please see the associated [design](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#appcontextbasedirectory) and [discussion](https://github.com/dotnet/designs/pull/90#discussion_r379218459).

Reply:
In that case, can you please add a comment into the code describing at least roughly what it does?
Do we think we will "get rid of it" in 6 - or deprecate it? I know we can't remove the bit from the struct, but maybe in 6 it would always be false or something. I would very much like to have a strategy how to get rid of it 😄

Reply:
We can certainly deprecate it in upcoming .net versions when: 
* A static host with linked runtime is available on all platforms, (for self-contained single-file apps). This is not available for Windows in .net5.
* With the upcoming corssgen2, we'll likely have cross platform native tools available for linking user's native dependencies.